File: server/src/main/java/com/yawa/server/validators/Username.java
package com.yawa.server.validators;

import javax.validation.Constraint;
import javax.validation.Payload;
import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.Target;

import static java.lang.annotation.ElementType.FIELD;
import static java.lang.annotation.RetentionPolicy.RUNTIME;

@Target({ FIELD })
@Retention(RUNTIME)
@Constraint(validatedBy = UsernameValidator.class)
@Documented
public @interface Username {

    String message() default UsernameValidator.VALIDATION_ERROR_MESSAGE;

    Class<?>[] groups() default { };

    Class<? extends Payload>[] payload() default { };

}

File: server/src/main/kotlin/com/yawa/server/validators/Username.kt
package com.yawa.server.validators

import java.util.regex.Pattern
import javax.validation.Constraint
import javax.validation.ConstraintValidator
import javax.validation.ConstraintValidatorContext
import javax.validation.Payload
import kotlin.reflect.KClass

@Target(AnnotationTarget.PROPERTY, AnnotationTarget.VALUE_PARAMETER)
@MustBeDocumented
@Constraint(validatedBy = [UsernameValidator::class])
annotation class Username(
    val message: String = "{com.yawa.server.validators.UsernameValidator.VALIDATION_ERROR_MESSAGE}",
    val groups: Array<KClass<*>> = [],
    val payload: Array<KClass<out Payload>> = []
)

class UsernameValidator : ConstraintValidator<Username, String> {

    val MIN_SIZE = 3
    val MAX_SIZE = 30
    val PATTERN : Pattern = Pattern.compile("^.{$MIN_SIZE,$MAX_SIZE}$")
    val VALIDATION_ERROR_MESSAGE = "Invalid username: must be a string between $MIN_SIZE and $MAX_SIZE characters"

    override fun isValid(value: String, context: ConstraintValidatorContext) : Boolean =
        PATTERN.matcher(value).matches()
}
